src/spatialdata_plot/pl/render.py

```diff
@@ -213,26 +230,64 @@ def _render_shapes(
+                cmap=ds_cmap,
                 color_key=color_key,
-                min_alpha=np.min([150, render_params.fill_alpha * 255]),
+                min_alpha=np.min([254, render_params.fill_alpha * 255]),
+                how="linear",
             )
-            if color_by_categorical or col_for_color is None
-            else ds.tf.shade(
+        elif aggregate_with_reduction is not None:  # to shut up mypy
+            ds_cmap = render_params.cmap_params.cmap
+            # in case all elements have the same value X: we render them using cmap(0.0),
+            # using an artificial "span" of [X, X + 1] for the color bar
+            # else: all elements would get alpha=0 and the color bar would have a weird range
+            if aggregate_with_reduction[0] == aggregate_with_reduction[1]:
+                ds_cmap = matplotlib.colors.to_hex(render_params.cmap_params.cmap(0.0), keep_alpha=False)
+                aggregate_with_reduction = (aggregate_with_reduction[0], aggregate_with_reduction[0] + 1)
+
+            ds_result = ds.tf.shade(
                 agg,
-                cmap=render_params.cmap_params.cmap,
+                cmap=ds_cmap,
+                how="linear",
+                min_alpha=np.min([254, render_params.fill_alpha * 255]),
+            )
+
+        # shade outlines if needed
+        outline_color = render_params.outline_params.outline_color
+        if isinstance(outline_color, str) and outline_color.startswith("#") and len(outline_color) == 9:
+            outline_color = outline_color[:-2]
+
+        if render_outlines:
+            ds_outlines = ds.tf.shade(
+                agg_outlines,
+                cmap=outline_color,
+                min_alpha=np.min([254, render_params.outline_alpha * 255]),
+                how="linear",
             )
+
+        rgba_image, trans_data = _create_image_from_datashader_result(ds_result, factor, ax)
+        _cax = _ax_show_and_transform(
+            rgba_image, trans_data, ax, zorder=render_params.zorder, alpha=render_params.fill_alpha
         )
-        rgba_image = np.transpose(ds_result.to_numpy().base, (0, 1, 2))
-        _cax = ax.imshow(rgba_image, cmap=palette, zorder=render_params.zorder)
-        _cax.set_transform(trans)
-        cax = ax.add_image(_cax)
-        if aggregate_with_sum is not None:
+        # render outline image if needed
+        if render_outlines:
+            rgba_image, trans_data = _create_image_from_datashader_result(ds_outlines, factor, ax)
+            _ax_show_and_transform(
+                rgba_image, trans_data, ax, zorder=render_params.zorder, alpha=render_params.outline_alpha
+            )
+
+        cax = None
+        if aggregate_with_reduction is not None:
             cax = ScalarMappable(
-                norm=matplotlib.colors.Normalize(vmin=aggregate_with_sum[0], vmax=aggregate_with_sum[1]),
+                norm=matplotlib.colors.Normalize(vmin=aggregate_with_reduction[0], vmax=aggregate_with_reduction[1]),
```

Comment:
Does this mean that a `Normalize` object passed on by the user will always be overwritten?

Reply:
That's a good point. I added that part so that the color map has the same range as the actual values. If you remove that I think that a `Normalize` object passed by the user right now has no effect at all (when using datashader), which is not ideal. I'll have a look into that

Reply:
Passing a `Normalize` object should work now. Not sure about the desired behavior of `clip` (see matplotlib example #372). @timtreis and I just discussed that we need more tests for this anyways, so I'm leaving it as is and we can do a follow-up